### ISSUE TITLE
Integrated gradients with LSTM to get attribution of combination of features 

### ISSUE THREAD

**Issue by sid-sundrani** [pytorch/captum#799]
## ❓ Questions and Help

I using Integrated gradients to obtain feature attributions within each timestep of an RNN. Using CAPTUM, Im able to get the attribution scores of each of the features at each timestep. Although, how would I get the attribution for a combination of features? Would I average the attribution scores? Or select min or max? 

For example, lets say I have an input matrix of (2 x 10 x 1 ) (seq len x num_features, batch_size) into a RNN performing sequence classification. Features 0-3 are features of a product's dimensions, 4-7 are features of a product name, and 8-10 are features of a products pricing. Using integrated gradients I can get the feature attributions of each of 10 features individually, but, what I want is only 3 feature importance attributions: product dimentions, product name, and product pricing. 

How would I go about getting this information? For example if I want the attribution of a products dimensions features, would I avg the attributions of 0-3 features?

**Comment by bilalsal** [comment]
Hi @sid-sundrani ,

apologies for responding late. 

If you are using gradient-based methods such as Integrated Gradients, then the solutions we currently support are along the same lines you thought, which is aggregating the individual feature attributions within each group. 

If you are using ablation-based methods such as [FeatureAblation](https://captum.ai/api/feature_ablation.html), then you can define a feature mask so that the features in the same group e.g. (0-3) are ablated together, and their attribution is effectively computed on the group level (Captum will calculate how much influence the group collectively has on the output).

[Here is a tutorial](https://captum.ai/tutorials/Resnet_TorchVision_Ablation) of `FeatureAblation` that demonstrates how you can define a feature mask.

Hope this helps

**Comment by damiaanr** [comment]
Excuse me for bumping this: what if one instead aims for an attribution score _per timestep_ using a similar architecture as OP? Would averaging the features within the temporal dimension be the way to go, when, for example, using IntegratedGradients?

**Comment by nazia-ga** [comment]
Hi, I am using Captum for my LSTM model. I followed the examples, but not sure I am using it the right way as I keep getting errors when starting to visualize it.  Please advise me on my code snippet below: 
```
# Define the LSTM model class
class LSTM(nn.Module):
    def __init__(self, input_size, hidden_size, num_stacked_layers):
        super().__init__()
        self.hidden_size = hidden_size
        self.num_stacked_layers = num_stacked_layers
        self.lstm = nn.LSTM(input_size, hidden_size, num_stacked_layers, batch_first=True)
        self.fc = nn.Linear(hidden_size, 1)

    def forward(self, x):
        batch_size = x.size(0)
        h0 = torch.zeros(self.num_stacked_layers, batch_size, self.hidden_size).to(device)
        c0 = torch.zeros(self.num_stacked_layers, batch_size, self.hidden_size).to(device)
        out, _ = self.lstm(x, (h0, c0))
        out = self.fc(out[:, -1, :])  # Output from last time step
        return out

model = LSTM(3,10,4)

train_losses = []
val_losses = []

def train_one_epoch():
    model.train(True)
    print(f'Epoch: {epoch + 1}')
    running_loss = 0.0

    for batch_index, batch in enumerate(train_loader):
        x_batch, y_batch = batch[0].to(device), batch[1].to(device)

        output = model(x_batch)
        loss = criterion(output, y_batch.view(-1, 1))
        running_loss += loss.item()
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

    avg_train_loss = running_loss / len(train_loader)
    train_losses.append(avg_train_loss)

def validate_one_epoch():
    model.eval()
    model.lstm.train()
    model.fc.train()

    running_loss = 0.0

    for batch_index, batch in enumerate(test_loader):
        x_batch, y_batch = batch[0].to(device), batch[1].to(device)

        with torch.no_grad():
            output = model(x_batch)
            loss = criterion(output, y_batch.view(-1, 1))
            running_loss += loss.item()
    
    avg_val_loss = running_loss / len(test_loader)
    val_losses.append(avg_val_loss)

learning_rate = 0.005
num_epochs = 2

criterion = nn.MSELoss()
optimizer = optim.Adam(model.parameters(), lr=learning_rate)

for epoch in range(num_epochs):
    print(f'Epoch {epoch + 1}/{num_epochs}')
    train_one_epoch()
    validate_one_epoch()
    print(f'Train Loss: {train_losses[-1]:.4f}, Val Loss: {val_losses[-1]:.4f}')
    print('*********************************')

 
with torch.no_grad():
    predicted = model(X_train_tensor.to(device)).to('cpu').numpy()

#Restore original scale for y
predicted_rescaled = scaler_y.inverse_transform(predicted)

# Similarly, inverse-transform y_train for comparison
y_train_rescaled = scaler_y.inverse_transform(y_train_scaled.reshape(-1, 1))

#Different attribution algorithms to the inputs of the model
ig = IntegratedGradients(model)
ig_nt = NoiseTunnel(ig)
dl = DeepLift(model)
gs = GradientShap(model)
fa = FeatureAblation(model)

ig_attr_test = ig.attribute(X_test_tensor, n_steps=50) #IntegratedGradients
ig_nt_attr_test = ig_nt.attribute(X_test_tensor) #NoiseTunnel
dl_attr_test = dl.attribute(X_test_tensor) #DeepLift
gs_attr_test = gs.attribute(X_test_tensor, X_train_tensor) #GradientShap
fa_attr_test = fa.attribute(X_test_tensor) #FeatureAblation

#for visualization
x_axis_data = np.arange(X_test_tensor.shape[1])
x_axis_data_labels = list(map(lambda idx: feature_names[idx], x_axis_data))

ig_attr_test_sum = ig_attr_test.detach().numpy().sum(0)
ig_attr_test_norm_sum = ig_attr_test_sum / np.linalg.norm(ig_attr_test_sum, ord=1)

ig_nt_attr_test_sum = ig_nt_attr_test.detach().numpy().sum(0)
ig_nt_attr_test_norm_sum = ig_nt_attr_test_sum / np.linalg.norm(ig_nt_attr_test_sum, ord=1)

dl_attr_test_sum = dl_attr_test.detach().numpy().sum(0)
dl_attr_test_norm_sum = dl_attr_test_sum / np.linalg.norm(dl_attr_test_sum, ord=1)

gs_attr_test_sum = gs_attr_test.detach().numpy().sum(0)
gs_attr_test_norm_sum = gs_attr_test_sum / np.linalg.norm(gs_attr_test_sum, ord=1)

fa_attr_test_sum = fa_attr_test.detach().numpy().sum(0)
fa_attr_test_norm_sum = fa_attr_test_sum / np.linalg.norm(fa_attr_test_sum, ord=1)

lin_weight = model.lin1.weight[0].detach().numpy()
y_axis_lin_weight = lin_weight / np.linalg.norm(lin_weight, ord=1)

width = 0.14
legends = ['Int Grads', 'Int Grads w/SmoothGrad','DeepLift', 'GradientSHAP', 'Feature Ablation', 'Weights']

plt.figure(figsize=(20, 10))

ax = plt.subplot()
ax.set_title('Comparing input feature importances across multiple algorithms and learned weights')
ax.set_ylabel('Attributions')

FONT_SIZE = 16
plt.rc('font', size=FONT_SIZE)            # fontsize of the text sizes
plt.rc('axes', titlesize=FONT_SIZE)       # fontsize of the axes title
plt.rc('axes', labelsize=FONT_SIZE)       # fontsize of the x and y labels
plt.rc('legend', fontsize=FONT_SIZE - 4)  # fontsize of the legend

ax.bar(x_axis_data, ig_attr_test_norm_sum, width, align='center', alpha=0.8, color='#eb5e7c')
ax.bar(x_axis_data + width, ig_nt_attr_test_norm_sum, width, align='center', alpha=0.7, color='#A90000')
ax.bar(x_axis_data + 2 * width, dl_attr_test_norm_sum, width, align='center', alpha=0.6, color='#34b8e0')
ax.bar(x_axis_data + 3 * width, gs_attr_test_norm_sum, width, align='center',  alpha=0.8, color='#4260f5')
ax.bar(x_axis_data + 4 * width, fa_attr_test_norm_sum, width, align='center', alpha=1.0, color='#49ba81')
ax.bar(x_axis_data + 5 * width, y_axis_lin_weight, width, align='center', alpha=1.0, color='grey')
ax.autoscale_view()
plt.tight_layout()

ax.set_xticks(x_axis_data + 0.5)
ax.set_xticklabels(x_axis_data_labels)

plt.legend(legends, loc=3)
plt.show()
```
